File: server/src/main/java/avatar/play/offline/AbsMapOffline.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package avatar.play.offline;

import avatar.model.Npc;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.Setter;

/**
 *
 * @author kitakeyos - Hoàng Hữu Dũng
 */
public abstract class AbsMapOffline {

    @Getter
    @Setter
    private int id;

    @Getter
    private List<Npc> npcs = new ArrayList<>();

    public AbsMapOffline(int id) {
        setId(id);
        init();
    }

    public abstract void init();

    public void addNpc(Npc npc) {
        synchronized (npcs) {
            npcs.add(npc);
        }
    }

    public void removeNpc(Npc npc) {
        synchronized (npcs) {
            npcs.remove(npc);
        }
    }
}

File: server-node/src/play/offline/AbsMapOffline.js
/**
 * Port của avatar/play/offline/AbsMapOffline.java
 * @author kitakeyos - Hoàng Hữu Dũng
 */
export class AbsMapOffline {

  constructor(id) {
    this.id = 0;
    this.npcs = [];
    this.setId(id);
    this.init();
  }

  /** abstract */
  init() {
    throw new Error('AbsMapOffline.init() is abstract');
  }

  addNpc(npc) {
    this.npcs.push(npc);
  }

  removeNpc(npc) {
    const idx = this.npcs.indexOf(npc);
    if (idx >= 0) {
      this.npcs.splice(idx, 1);
    }
  }

  // ==== getter/setter (lombok) ====
  getId() { return this.id; }
  setId(v) { this.id = v | 0; }

  getNpcs() { return this.npcs; }
}

export default AbsMapOffline;
